redisson/src/main/java/org/redisson/RedissonPermitExpirableSemaphore.java

```diff
@@ -645,11 +655,79 @@ public RFuture<Integer> availablePermitsAsync() {
+                    "if tonumber(value) > 0 then " +
+                        "redis.call('publish', KEYS[3], value); " +
+                    "end;" +
+                "end; " +
+                "local available = redis.call('get', KEYS[1]); " +
+                "if available == false then " +
+                    "return 0 " +
+                "end;" +
+                "local claimed = redis.call('zcount', KEYS[2], 0, '+inf'); " +
+                "if claimed == false then " +
+                    "return tonumber(available) " +
+                "end;" +
+                "return tonumber(available) + claimed;",
+                Arrays.<Object>asList(getRawName(), timeoutName, getChannelName()), System.currentTimeMillis());
+    }
+
+    @Override
+    public RFuture<Integer> claimedPermitsAsync() {
+        return commandExecutor.evalWriteAsync(getRawName(), LongCodec.INSTANCE, RedisCommands.EVAL_INTEGER,
+                "local expiredIds = redis.call('zrangebyscore', KEYS[2], 0, ARGV[1], 'limit', 0, -1); " +
+                "if #expiredIds > 0 then " +
+                    "redis.call('zrem', KEYS[2], unpack(expiredIds)); " +
+                    "local value = redis.call('incrby', KEYS[1], #expiredIds); " +
+                    "if tonumber(value) > 0 then " +
+                        "redis.call('publish', KEYS[3], value); " +
+                    "end;" +
+                "end; " +
+                "local claimed = redis.call('zcount', KEYS[2], 0, '+inf'); " +
+                "return claimed == false and 0 or claimed;",
+                Arrays.<Object>asList(getRawName(), timeoutName, getChannelName()), System.currentTimeMillis());
+    }
+
     @Override
     public boolean trySetPermits(int permits) {
         return get(trySetPermitsAsync(permits));
     }
 
+    @Override
+    public int setPermits(int permits) {
+        return get(setPermitsAsync(permits));
+    }
+
+    @Override
+    public RFuture<Integer> setPermitsAsync(int permits) {
+        return commandExecutor.evalWriteAsync(getRawName(), LongCodec.INSTANCE, RedisCommands.EVAL_INTEGER,
+                "local available = redis.call('get', KEYS[1]); " +
+                "if (available == false) then " +
+                    "redis.call('set', KEYS[1], ARGV[1]); " +
+                    "redis.call('publish', KEYS[2], ARGV[1]); " +
+                    "return ARGV[1];" +
+                "end;" +
+                "local claimed = redis.call('zcount', KEYS[3], 0, '+inf'); " +
+                "local maximum = (claimed == false and 0 or claimed) + tonumber(available); " +
+                "if (maximum == ARGV[1]) then " +
+                    "return 0;" +
+                "end;" +
+                "local delta = tonumber(ARGV[1]) - maximum; " +
+                "if (delta == 0) then " +
+                    "return 0;" +
+                "end;" +
+                "redis.call('incrby', KEYS[1], delta); " +
```

Comment:
what if I set permits less than before. Let's say, currently semaphore has 10 permits then we set 5. I don't see how this case handled in this code. Delta will be negative in this case.

Reply:
`incrby` handles a negative delta as a decrement so it will be decremented to match the existing behaviour of addPermits. 

There are a few test cases which walk through decreasing the number of permits using both the new [setPermits](https://github.com/redisson/redisson/pull/4719/files#diff-8d51b6c35753a38648e0759011ae9ebb83d954359ba65801f5cd117a647ea5f9R227-R238) as well as [addPermits](https://github.com/redisson/redisson/pull/4719/files#diff-8d51b6c35753a38648e0759011ae9ebb83d954359ba65801f5cd117a647ea5f9R248-R251)

Reply:
availablePermits() method will return negative number which is incorrect.

Reply:
Right - thats the behaviour today as well when you use `addPermits` with a negative number that makes the number of permits below zero. I could update the `availablePermits` so that it doesn't return a negative number.

There are valid cases where the number of permits could below zero until some clients release permits. If there are 10 permits and clients have acquired all 10 of them, then we set the number of permits to 5, then the available permits would be negative five, and would not get back to a positive number until 6 clients have released permits.